docs/design/generics/details.md

```diff
@@ -2463,549 +2473,1280 @@ interface A(T:! B where ...) {
-have a function only accept stacks containing integers:
+  let SliceType:! Container
+      where .ElementType = ElementType and
+            // `.Self` means `SliceType`.
+            .SliceType = .Self;
 
-```
-fn SumIntStack[T:! Stack where .ElementType = i32](s: T*) -> i32 {
-  var sum: i32 = 0;
-  while (!s->IsEmpty()) {
-    // s->Pop() has type `T.ElementType` == i32:
-    sum += s->Pop();
-  }
-  return sum;
+  // `Self` means the type implementing `Container`.
+  fn GetSlice[addr self: Self*]
+      (start: IteratorType, end: IteratorType) -> SliceType;
 }
 ```
 
-To name these sorts of constraints, we could use `let` declarations or
-`constraint` definitions:
+Note that [naming](#named-constraint-constants) a recursive constraint using the
+[`constraint` introducer](#named-constraints) approach, we can name the
+implementing type using `Self` instead of `.Self`, since they refer to the same
+type:
 
-```
-let IntStack:! auto = Stack where .ElementType = i32;
-constraint IntStack {
-  extend Stack where .ElementType = i32;
+```carbon
+constraint RealAbs {
+  extend HasAbs where .MagnitudeType = Self;
+  // Equivalent to:
+  extend HasAbs where .MagnitudeType = .Self;
+}
+
+constraint ContainerIsSlice {
+  extend Container where .SliceType = Self;
+  // Equivalent to:
+  extend Container where .SliceType = .Self;
 }
 ```
 
-This syntax is also used to specify the values of
-[associated constants](#associated-constants) when implementing an interface for
-a type.
+The `.Self` construct follows these rules:
 
-##### Equal generic types
+-   `X :!` introduces `.Self:! type`, where references to `.Self` are resolved.
+    to `X`. This allows you to use `.Self` as an interface parameter as in
+    `X:! I(.Self)`.
+-   `A where` introduces `.Self:! A` and `.Foo` _designator_ for each member
+    `Foo` of `A`.
+-   It's an error to reference `.Self` if it refers to more than one different
+    thing or isn't a facet.
+-   You get the innermost, most-specific type for `.Self` if it is introduced
+    twice in a scope. By the previous rule, it is only legal if they all refer
+    to the same facet binding.
+-   `.Self` may not be on the left side of the `=` in a rewrite constraint.
 
-Alternatively, two generic types could be constrained to be equal to each other,
-without specifying what that type is. This uses `==` instead of `=`. For
-example, we could make the `ElementType` of an `Iterator` interface equal to the
-`ElementType` of a `Container` interface as follows:
+So in `X:! A where ...`, `.Self` is introduced twice, after the `:!` and the
+`where`. This is allowed since both times it means `X`. After the `:!`, `.Self`
+has the type `type`, which gets refined to `A` after the `where`. In contrast,
+it is an error if `.Self` could mean two different things, as in:
 
+```carbon
+// ❌ Illegal: `.Self` could mean `T` or `T.A`.
+fn F[T:! InterfaceA where .A impls
+           (InterfaceB where .B = .Self)](x: T);
```

Comment:
Is it interesting to say how to write the two meanings? I think they're:
```
constraint InterfaceBWithSelf {
  extend InterfaceB where .B = Self;
}
constraint InterfaceBWith(T:! InterfaceA) {
  extend InterfaceB where .B = T;
}
fn F[T:! InterfaceA where .A impls InterfaceBWithSelf](x: T);
fn F[T:! InterfaceA where .A impls InterfaceBWith(.Self)](x: T);
```

Reply:
Added.